New API Attach/Detach for Virtual Machines and Virtual Machines Scaleset

csassasin opened the pull request:
## ARM API Information (Control Plane)

The Control Plane is planning to come up with the below features:-

> New API for Attach/Detach of data disks to/from a VirtualMachine.
> New API for Attach/Detach of data disks to/from a VirtualMachineScaleSet(Uniform)'s VM.


 


### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [ ] new API version
      - [x] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [x] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version). Note that this doesn't apply if you are trying to merge a PR that was previously in the private repository.

- [x] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/overview_breakingchanges).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](http://aka.ms/aprwiki))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

psuradms committed:
Swagger changes for new Attach/Detach data disks APIs

- New API for Attach/Detach of data disks to/from a VirtualMachine
- New API for Attach/Detach of data disks to/from a VirtualMachineScaleSet(Uniform)'s VM

psuradms committed:
Incorporating review comments

psuradms committed:
Incorporating review comments

Commit:
Detached DetachedDataDisk property added

Detached DetachedDataDisk property added in response apart from attachDataDisk.

ms-zhenhua commented on specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2023-07-01/virtualMachine.json:
```diff
@@ -1891,6 +1891,72 @@
+            "required": true,
+            "type": "string",
+            "description": "The name of the virtual machine."
+          },
```
should be defined with a 'pattern' 
[https://github.com/Azure/azure-openapi-validator/blob/main/docs/resource-name-restriction.md](https://github.com/Azure/azure-openapi-validator/blob/main/docs/resource-name-restriction.md)

csassasin replied:
@ms-zhenhua Adding the pattern here. But to be clear currently the resourceName doesn't contains pattern. I hope this would not create problems for existing resource names

ms-zhenhua commented on specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2023-07-01/virtualMachine.json:
```diff
@@ -1891,6 +1891,72 @@
+            "$ref": "../../../common-types/v1/common.json#/parameters/SubscriptionIdParameter"
+          }
+        ],
+        "responses": {
```
>responses [](http://example.com/codeflow?start=9&length=9)

should have 200:
[https://github.com/Azure/azure-openapi-validator/blob/main/docs/lro-post-return.md](https://github.com/Azure/azure-openapi-validator/blob/main/docs/lro-post-return.md)

ms-zhenhua commented on specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2023-07-01/virtualMachine.json:
```diff
@@ -1891,6 +1891,72 @@
+          "202": {
+            "description": "Accepted",
+            "schema": {
+              "$ref": "./computeRPCommon.json#/definitions/AttachDetachDataDisksResponse"
```
>"$ref": "./computeRPCommon.json#/definitions/AttachDetachDataDisksResponse" [](http://example.com/codeflow?start=14&length=75)

202 should not have schema:
[https://github.com/Azure/azure-openapi-validator/blob/main/docs/lro-post-return.md](https://github.com/Azure/azure-openapi-validator/blob/main/docs/lro-post-return.md)

csassasin replied:
Changes made. Currently based on team discussion with team:
Finalized HTTP Verb: HTTP POST
Response will contain both 200 and 202.
In 200, response "azure-asyncoperation" header will be provided. This is similar to start operation on the VM https://learn.microsoft.com/en-us/rest/api/compute/virtual-machines/start?tabs=HTTP#virtualmachine_start_maximumset_gen

ms-zhenhua commented on specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2023-07-01/virtualMachine.json:
```diff
@@ -1891,6 +1891,72 @@
+          }
+        ],
+        "responses": {
+          "202": {
```
>202 [](http://example.com/codeflow?start=11&length=3)

202 should have location header:
[https://github.com/Azure/azure-openapi-validator/blob/main/docs/lro-location-header.md](https://github.com/Azure/azure-openapi-validator/blob/main/docs/lro-location-header.md)

csassasin replied:
Changes made. Currently based on team discussion with team:
Finalized HTTP Verb: HTTP POST
Response will contain both 200 and 202.
In 200, response "azure-asyncoperation" header will be provided. This is similar to start operation on the VM https://learn.microsoft.com/en-us/rest/api/compute/virtual-machines/start?tabs=HTTP#virtualmachine_start_maximumset_gen

ms-zhenhua commented on specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2023-07-01/virtualMachineScaleSet.json:
```diff
@@ -3489,6 +3489,79 @@
+        "responses": {
+          "202": {
+            "description": "Accepted",
+            "schema": {
```
>schema [](http://example.com/codeflow?start=13&length=6)

202 must not have schema:
[https://github.com/Azure/azure-openapi-validator/blob/main/docs/lro-post-return.md](https://github.com/Azure/azure-openapi-validator/blob/main/docs/lro-post-return.md)

csassasin replied:
Adhering to this . 
Currently based on team discussion with team: 
Finalized HTTP Verb: HTTP POST
Response will contain both 200 and 202.
In 200, response "azure-asyncoperation" header will be provided. This is similar to start operation on the VM https://learn.microsoft.com/en-us/rest/api/compute/virtual-machines/start?tabs=HTTP#virtualmachine_start_maximumset_gen

ms-zhenhua commented on specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2023-07-01/virtualMachineScaleSet.json:
```diff
@@ -3489,6 +3489,79 @@
+          }
+        ],
+        "responses": {
+          "202": {
```
>202 [](http://example.com/codeflow?start=11&length=3)

202 must have location header:
[https://github.com/Azure/azure-openapi-validator/blob/main/docs/lro-location-header.md](https://github.com/Azure/azure-openapi-validator/blob/main/docs/lro-location-header.md)

csassasin replied:
Changes made. Currently based on team discussion with team:
Finalized HTTP Verb: HTTP POST
Response will contain both 200 and 202.
In 200, response "azure-asyncoperation" header will be provided. This is similar to start operation on the VM https://learn.microsoft.com/en-us/rest/api/compute/virtual-machines/start?tabs=HTTP#virtualmachine_start_maximumset_gen

ms-zhenhua commented on specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2023-07-01/virtualMachineScaleSet.json:
```diff
@@ -3489,6 +3489,79 @@
+          "default": {
+            "description": "Error response describing why the operation failed.",
+            "schema": {
+              "$ref": "../../../common-types/v1/common.json#/definitions/CloudError"
```
>"../../../common-types/v1/common.json#/definitions/CloudError" [](http://example.com/codeflow?start=22&length=62)

use common type:
[https://github.com/Azure/azure-openapi-validator/blob/main/docs/lro-error-content.md](https://github.com/Azure/azure-openapi-validator/blob/main/docs/lro-error-content.md)

csassasin replied:
Same as older API ways of handling it with version v1 in Microsoft.Compute. Resolving this , please feel free to open in case it requires discussion

ms-zhenhua commented on specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2023-07-01/virtualMachineScaleSet.json:
```diff
@@ -3489,6 +3489,79 @@
+            "required": true,
+            "type": "string",
+            "description": "The name of the resource group."
+          },
```
use the latest "v5" common type:

[https://github.com/Azure/azure-openapi-validator/blob/main/docs/parameter-not-using-common-types.md](https://github.com/Azure/azure-openapi-validator/blob/main/docs/parameter-not-using-common-types.md)

csassasin replied:
@ms-zhenhua Currently in Microsoft.Compute I cannot see the version v5 specified anywhere . I can see even in main we are in v1 version. Can you please elaborate how v5 can be specified .

ms-zhenhua commented on specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2023-07-01/virtualMachine.json:
```diff
@@ -1891,6 +1891,72 @@
+            "required": true,
+            "type": "string",
+            "description": "The name of the resource group."
+          },
```
use the latest "v5" common type:

[https://github.com/Azure/azure-openapi-validator/blob/main/docs/parameter-not-using-common-types.md](https://github.com/Azure/azure-openapi-validator/blob/main/docs/parameter-not-using-common-types.md)

csassasin replied:
@ms-zhenhua  Currently in Microsoft.Compute I cannot see the version v5 specified anywher . I can see even in main we are in v1 version. Can you please elaborate how v5 can be specified. Not clear with the ask here

ms-zhenhua commented on specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2023-07-01/computeRPCommon.json:
```diff
@@ -1913,6 +1913,103 @@
+    },
+    "AttachDataDisk": {
+      "properties": {
+        "id": {
```
>id [](http://example.com/codeflow?start=9&length=2)

"id" is used in top-level model, should not be used in nested properties, please change the name.

csassasin replied:
Changed to diskId . Please point out the id in the top-level model for better understanding

ms-zhenhua commented on specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2023-07-01/computeRPCommon.json:
```diff
@@ -1913,6 +1913,103 @@
+    },
+    "DetachDataDisk": {
+      "properties": {
+        "id": {
```
>id [](http://example.com/codeflow?start=9&length=2)

"id" is used in top-level model, should not be used in nested properties, please change the name.

csassasin replied:
Changed to diskId . Please point out the id in the top-level model for better understanding

ms-zhenhua commented on specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2023-07-01/computeRPCommon.json:
```diff
@@ -1913,6 +1913,103 @@
+    },
+    "AttachedDataDisk": {
+      "properties": {
+        "id": {
```
>id [](http://example.com/codeflow?start=9&length=2)

"id" is used in top-level model, should not be used in nested properties, please change the name.

csassasin replied:
Changed to diskId . Please point out the id in the top-level model for better understanding

ms-zhenhua commented on specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2023-07-01/virtualMachine.json:
```diff
@@ -1891,6 +1891,72 @@
+          },
+          {
+            "$ref": "../../../common-types/v1/common.json#/parameters/SubscriptionIdParameter"
+          }
```
put it to the first to make the definition consistent with the url

csassasin replied:
Modified thanks for the suggestion.

csassasin commented on specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2023-07-01/computeRPCommon.json:
```diff
@@ -1913,6 +1913,103 @@
+          "readOnly": true,
+          "description": "ID of the managed data disk."
+        },
+        "lun": {
```
Sticking to the older design for Virtual Machines rest semantics: https://learn.microsoft.com/en-us/rest/api/compute/virtual-machines/create-or-update?tabs=HTTP

csassasin commented on specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2023-07-01/virtualMachine.json:
```diff
@@ -1891,6 +1891,72 @@
+            "description": "Parameters supplied to the attach and detach data disks operation on the virtual machine."
+          },
+          {
+            "$ref": "../../../common-types/v1/common.json#/parameters/ApiVersionParameter"
```
@ms-zhenhua Currently in Microsoft.Compute I cannot see the version v5 specified anywher . I can see even in main we are in v1 version. Can you please elaborate how v5 can be specified. Not clear with the ask here

csassasin replied:
Resolving it based on offline discussion as Microsoft.Compute is still on version v1

csassasin commented on specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2023-07-01/virtualMachine.json:
```diff
@@ -1891,6 +1891,72 @@
+          "default": {
+            "description": "Error response describing why the operation failed.",
+            "schema": {
+              "$ref": "../../../common-types/v1/common.json#/definitions/CloudError"
```
@ms-zhenhua Currently in Microsoft.Compute I cannot see the version v5 specified anywhere . I can see even in main we are in v1 version. Can you please elaborate how v5 can be specified .

csassasin commented on specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2023-07-01/virtualMachineScaleSet.json:
```diff
@@ -3489,6 +3489,79 @@
+            "$ref": "../../../common-types/v1/common.json#/parameters/ApiVersionParameter"
+          },
+          {
+            "$ref": "../../../common-types/v1/common.json#/parameters/SubscriptionIdParameter"
```
Thanks for the suggestion. Changes being made.

Commit:
Reordering of the references have been changed

Reordering of the references have been changed

Commit:
Updated "id" with diskId based on review comments by ARM .

Updated "id" with diskId based on review comments by ARM .

csassasin commented on specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2023-07-01/computeRPCommon.json:
```diff
@@ -1913,6 +1913,103 @@
+    },
+    "DetachedDataDisk": {
+      "properties": {
+        "id": {
```
changes to diskId . Just for reference please point out where is ID being used in top-level model.

Commit:
Reverting to 202 , post sync up with internal team discussion will fix this

Commit:
Review comments regarding 202 and 200 addressed

For the newer API
200 : Its represents the operation is OK . Customer can poll on the operation using the header
202: Request is accepted

csassasin commented on specification/compute/resource-manager/Microsoft.Compute/ComputeRP/stable/2023-07-01/virtualMachineScaleSet.json:
```diff
@@ -3489,6 +3489,79 @@
+            "$ref": "../../../common-types/v1/common.json#/parameters/SubscriptionIdParameter"
+          }
+        ],
+        "responses": {
```
Changes made. Currently based on team discussion with team:
Finalized HTTP Verb: HTTP POST
Response will contain both 200 and 202.
In 200, response "azure-asyncoperation" header will be provided. This is similar to start operation on the VM https://learn.microsoft.com/en-us/rest/api/compute/virtual-machines/start?tabs=HTTP#virtualmachine_start_maximumset_gen

mentat9 commented:
@cassasin - Changing response codes requires breaking change review (see comment for instructions: https://github.com/Azure/azure-rest-api-specs/pull/24334#issuecomment-1594510134).

Commit:
Updated virtualMachines.json

Commit:
Updated examples

raosuhas commented:
Can you please fix the LintDiff error introduced by this PR here : https://github.com/Azure/azure-rest-api-specs/pull/24334/checks?check_run_id=14934248097 .

raosuhas commented:
> Can you please fix the LintDiff error introduced by this PR here : https://github.com/Azure/azure-rest-api-specs/pull/24334/checks?check_run_id=14934248097 .

If you are unable to fix these for any reason. Please add suppressions along with reasons so that it can be reviewed

csassasin commented:
> > Can you please fix the LintDiff error introduced by this PR here : https://github.com/Azure/azure-rest-api-specs/pull/24334/checks?check_run_id=14934248097 .
> 
> If you are unable to fix these for any reason. Please add suppressions along with reasons so that it can be reviewed

The lint errors are same as the previous ones. Those cannot be fixed because that would start breaking the existing API's.  
![image](https://github.com/Azure/azure-rest-api-specs/assets/89677872/8fb5c6b7-e04e-4649-a3b5-66881f36cae1)
